tremor-codec/src/codec/graphite.rs

```diff
@@ -0,0 +1,243 @@
+        m.insert("timestamp".into(), Value::from(ts));
+    }
+
+    m.insert("value".into(), value);
+    Ok(Value::from(m))
+}
+
+#[cfg(test)]
+mod test {
+
+    mod plaintext {
+        use super::super::*;
+        use tremor_value::literal;
+
+        // beep.boop 7 1620649445
+        #[tokio::test(flavor = "multi_thread")]
+        async fn name_value_ts() {
+            let data = b"beep.boop 7 1620649445";
+            let parsed = decode_plaintext(data, 0).expect("failed to decode");
+            let expected = literal!({
+                "metric": "beep.boop",
+                "value": 7,
+                "timestamp": 1_620_649_445i64,
+
+            });
+            assert_eq!(parsed, expected);
+            let mut encoded = Vec::new();
+            encode_plaintext(&parsed, &mut encoded).expect("failed to encode");
+            assert_eq!(encoded.as_slice(), data);
+        }
+
+        // beep.boop 320.0 -1
+        #[tokio::test(flavor = "multi_thread")]
+        async fn name_value_minusone() {
+            let data = b"beep.boop 320.0 -1";
+            let parsed = decode_plaintext(data, 1234).expect("failed to decode");
+            let expected = literal!({
+                "metric": "beep.boop",
+                "value": 320.0,
+                "timestamp": 1234,
+            });
+            assert_eq!(parsed, expected);
+            let mut encoded = Vec::new();
+            encode_plaintext(&parsed, &mut encoded).expect("failed to encode");
+            assert_eq!(encoded, b"beep.boop 320.0 1234"); // -1 is replaced with 1234 so this is asymetric w.r.t. input
+        }
+
+        #[tokio::test(flavor = "multi_thread")]
+        async fn test_for_coverage() {
+            let mut codec = PlaintextProtocol::default();
+            assert_eq!("graphite-plaintext", codec.name());
+            let cloned = codec.boxed_clone();
+            assert_eq!("graphite-plaintext", cloned.name());
+
+            let mut data = b"beep.boop 7 1620649445".to_vec();
+            let decoded = codec.decode(&mut data, 0, Value::null()).await;
+            assert_eq!(
+                decoded,
+                Ok(Some((
+                    literal!({
+                        "metric": "beep.boop",
+                        "value": 7,
+                        "timestamp": 1_620_649_445i64,
+                    }),
+                    Value::null()
+                )))
+            );
+
+            let mut bad = b"".to_vec();
+            let decoded = codec.decode(&mut bad, 0, Value::null()).await;
+            assert!(decoded.is_err());
+
+            let mut bad = b"metric.no.measure.no.timestamp".to_vec();
+            let decoded = codec.decode(&mut bad, 0, Value::null()).await;
+            assert!(decoded.is_err());
+
+            let mut bad = b"metric.with.measure.no.timestamp 320.0".to_vec();
+            let decoded = codec.decode(&mut bad, 0, Value::null()).await;
+            assert!(decoded.is_err());
```

Comment:
Based on the documentation this should parse

> //! If timestamps are not provided, the codec will use the current time as the event's timestamp.

Reply:
I'll have to reread and check as there's precious little. You might be right though assuming it's the same as -1 and gets current time

Reply:
Ja, at least from the docs I'd expect no timestamp acting the same as -1

Reply:
I think that's another vestigial thing as I used statsd as a base to get things close. The graphite docs don't have this
behaviour though at all from my reading. So the comment was false, but the test correct. I should have removed the comment when I added the test:

https://graphite.readthedocs.io/en/latest/feeding-carbon.html#step-3-understanding-the-graphite-message-format

I'll update accordingly

Reply:
Agreed, the docs were off the test + code are <3

Reply:
Pushing now. Well spotted either way!